### PR TITLE
privilege/privileges: sort in-memory db privilege table records

### LINKED ISSUE #37488: Access control and privileges check use diffrent host when USER table has same username
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1.
mysql -uroot -p -h192.168.13.15 -P4000
create user dba_test@’%’ identified by ‘123456’;
GRANT SELECT,INSERT,UPDATE,DELETE ON test.* TO ‘dba_test’@’%’;
create user dba_test@‘192.168.%’ identified by ‘654321’;
GRANT SELECT,INSERT,UPDATE,DELETE,CREATE,DROP,ALTER ON test.* TO ‘dba_test’@‘192.168.%’;

2.
mysql -udba_test -p654321 -h192.168.13.15 -P4000
MySQL [(none)]> use test;
Database changed
MySQL [test]> DROP TABLE IF EXISTS a;
ERROR 1142 (42000): DROP command denied to user ‘dba_test’@‘192.168.%’ for table ‘a’

3.
mysql -uroot -p -h192.168.13.15 -P4000
MySQL [(none)]> drop user dba_test@'%';
Query OK, 0 rows affected (0.23 sec)

4.
mysql -udba_test -p654321 -h192.168.13.15 -P4000
MySQL [(none)]> use test;
Database changed
MySQL [test]> DROP TABLE IF EXISTS a;
Query OK, 0 rows affected, 1 warning (0.01 sec)

### 2. What did you expect to see? (Required)
MySQL [test]> DROP TABLE IF EXISTS a;
Query OK, 0 rows affected, 1 warning (0.01 sec)

### 3. What did you see instead (Required)
MySQL [test]> DROP TABLE IF EXISTS a;
ERROR 1142 (42000): DROP command denied to user ‘dba_test’@‘192.168.%’ for table ‘a’

### 4. What is your TiDB version? (Required)
v5.4.1,v6.1.0

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #37488

Problem Summary:

> It is possible for the client host name and user name of an incoming connection to match more than one row in the user table.

According to [MySQL document](https://dev.mysql.com/doc/refman/8.0/en/connection-access.html), when multiple matches are possible, the server must determine which of them to use by some sorting rules.

We've handle the sorting in the past for 'mysql.user' records, but not for 'mysql.db', and this cause the bug of #37488

### What is changed and how it works?

Sort the loaded 'mysql.db' records, so the privileges checker can match the correct record and handle the multiple-match case.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: privilege/privileges/cache.go [hunk 1]
```diff
@@ -535,6 +536,14 @@ func (p *MySQLPrivilege) buildDBMap() {
 	for _, record := range p.DB {
 		dbMap[record.User] = append(dbMap[record.User], record)
 	}
+
+	// Sort the records to make the matching rule work.
+	for _, records := range dbMap {
+		sort.Slice(records, func(i, j int) bool {
```
Comment (hawkingrei) [inline privilege/privileges/cache.go:542]: if you rewrite  ```compareBaseRecord```  as follows：
```
func compareBaseRecord(x, y baseRecord) bool
```

you can use ```slice.SortFunc``` to replace ```sort.Slice```.

```
slice.SortFunc(records, compareBaseRecord)
```

it is more simple and avoids BCE in the Less function.

Comment (tiancaiamao) [reply privilege/privileges/cache.go]: `compareBaseRecord` can't work.

```
cat aaa.go
package main

import (
	"golang.org/x/exp/slices"
)

type baseT struct {
	int
}

type Extend struct {
	baseT
}

func compareBaseT(x, y baseT) bool {
	return x.int < y.int
}

func main() {
	xx := []Extend{}
	slices.SortFunc(xx, compareBaseT)
}
```

```
go run aaa.go
# command-line-arguments
./aaa.go:21:22: type func(x baseT, y baseT) bool of compareBaseT does not match inferred type func(a Extend, b Extend) bool for func(a E, b E) bool
```

It need a `compareDBRecord` function

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**